lib/Service/BreakoutRoomService.php

```diff
@@ -0,0 +1,237 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace OCA\Talk\Service;
+
+use InvalidArgumentException;
+use OCA\Talk\Manager;
+use OCA\Talk\Model\Attendee;
+use OCA\Talk\Model\BreakoutRoom;
+use OCA\Talk\Participant;
+use OCA\Talk\Room;
+use OCA\Talk\Webinary;
+use OCP\EventDispatcher\IEventDispatcher;
+use OCP\IL10N;
+
+class BreakoutRoomService {
+	protected Manager $manager;
+	protected RoomService $roomService;
+	protected ParticipantService $participantService;
+	protected IEventDispatcher $dispatcher;
+	protected IL10N $l;
+
+	public function __construct(Manager $manager,
+								RoomService $roomService,
+								ParticipantService $participantService,
+								IEventDispatcher $dispatcher,
+								IL10N $l) {
+		$this->manager = $manager;
+		$this->roomService = $roomService;
+		$this->participantService = $participantService;
+		$this->dispatcher = $dispatcher;
+		$this->l = $l;
+	}
+
+	/**
+	 * @param Room $parent
+	 * @param int $mode
+	 * @psalm-param 0|1|2|3 $mode
+	 * @param int $amount
+	 * @param string $attendeeMap
+	 * @return Room[]
+	 * @throws InvalidArgumentException When the breakout rooms are configured already
+	 */
+	public function setupBreakoutRooms(Room $parent, int $mode, int $amount, string $attendeeMap): array {
+		if ($parent->getBreakoutRoomMode() !== BreakoutRoom::MODE_NOT_CONFIGURED) {
+			throw new InvalidArgumentException('room');
+		}
+
+		if ($parent->getType() !== Room::TYPE_GROUP
+			&& $parent->getType() !== Room::TYPE_PUBLIC) {
+			// Can only do breakout rooms in group and public rooms
+			throw new InvalidArgumentException('room');
+		}
+
+		if ($parent->getObjectType() === BreakoutRoom::PARENT_OBJECT_TYPE) {
+			// Can not nest breakout rooms
+			throw new InvalidArgumentException('room');
+		}
+
+		if (!$this->roomService->setBreakoutRoomMode($parent, $mode)) {
+			throw new InvalidArgumentException('mode');
+		}
+
+		if ($amount < BreakoutRoom::MINIMUM_ROOM_AMOUNT) {
+			throw new InvalidArgumentException('amount');
+		}
+
+		if ($amount > BreakoutRoom::MAXIMUM_ROOM_AMOUNT) {
+			throw new InvalidArgumentException('amount');
+		}
+
+		if ($mode === BreakoutRoom::MODE_MANUAL) {
+			try {
+				$attendeeMap = json_decode($attendeeMap, true, 2, JSON_THROW_ON_ERROR);
+			} catch (\JsonException $e) {
+				throw new InvalidArgumentException('attendeeMap');
+			}
+
+			if (!empty($attendeeMap)) {
+				if (max($attendeeMap) >= $amount) {
+					throw new InvalidArgumentException('attendeeMap');
+				}
+
+				if (min($attendeeMap) < 0) {
+					throw new InvalidArgumentException('attendeeMap');
+				}
+			}
+		}
+
+		$breakoutRooms = $this->createBreakoutRooms($parent, $amount);
+
+		$participants = $this->participantService->getParticipantsForRoom($parent);
```

Comment:
I think that will be good to check if all members of `$attendeeMap` are participants.

Reply:
Only to prevent to send in this array an user that isn't member of room.